File: vending-app/src/main/java/com/android/vending/licensing/LicensingService.java
/*
 * SPDX-FileCopyrightText: 2023 microG Project Team
 * SPDX-License-Identifier: Apache-2.0
 */

package com.android.vending.licensing;

import static com.android.vending.licensing.LicenseChecker.LICENSED;

import android.accounts.Account;
import android.accounts.AccountManager;
import android.app.Service;
import android.content.Intent;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.os.Bundle;
import android.os.IBinder;
import android.os.RemoteException;
import android.net.Uri;
import android.net.Uri;
import android.os.Bundle;
import android.os.IBinder;
import android.os.RemoteException;
import android.util.Log;

import com.android.vending.VendingPreferences;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.Volley;

import org.microg.gms.auth.AuthConstants;
import org.microg.gms.profile.ProfileManager;

import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.Queue;

import kotlin.Unit;

public class LicensingService extends Service {
    private static final String TAG = "FakeLicenseService";
    private RequestQueue queue;
    private AccountManager accountManager;
    private LicenseServiceNotificationRunnable notificationRunnable;
    private String androidId;

    private static final String KEY_V2_RESULT_JWT = "LICENSE_DATA";

    private static final Uri PROFILE_PROVIDER = Uri.parse("content://com.google.android.gms.microg.profile");

    private static final Uri CHECKIN_SETTINGS_PROVIDER = Uri.parse("content://com.google.android.gms.microg.settings/check-in");

    private final ILicensingService.Stub mLicenseService = new ILicensingService.Stub() {

        @Override
        public void checkLicense(long nonce, String packageName, ILicenseResultListener listener) throws RemoteException {
            Log.v(TAG, "checkLicense(" + nonce + ", " + packageName + ")");
            int callingUid = getCallingUid();

            if (!VendingPreferences.isLicensingEnabled(getApplicationContext())) {
                Log.d(TAG, "not checking license, as it is disabled by user");
                return;
            }

            Account[] accounts = accountManager.getAccountsByType(AuthConstants.DEFAULT_ACCOUNT_TYPE);
            PackageManager packageManager = getPackageManager();

            if (accounts.length == 0) {
                handleNoAccounts(packageName, packageManager);
            } else {
                checkLicense(callingUid, nonce, packageName, packageManager, listener, new LinkedList<>(Arrays.asList(accounts)));
            }
        }

        private void checkLicense(int callingUid, long nonce, String packageName, PackageManager packageManager,
                                  ILicenseResultListener listener, Queue<Account> remainingAccounts) throws RemoteException {
            new LicenseChecker.V1().checkLicense(
                remainingAccounts.poll(), accountManager, androidId, packageName, callingUid, packageManager,
                queue, nonce,
                (responseCode, stringTuple) -> {
                    if (responseCode != LICENSED && !remainingAccounts.isEmpty()) {
                        checkLicense(callingUid, nonce, packageName, packageManager, listener, remainingAccounts);
                    } else {
                        listener.verifyLicense(responseCode, stringTuple != null ? stringTuple.a : null,  stringTuple != null ? stringTuple.b : null);
                    }
                }
            );
        }

        @Override
        public void checkLicenseV2(String packageName, ILicenseV2ResultListener listener, Bundle extraParams) throws RemoteException {
            Log.v(TAG, "checkLicenseV2(" + packageName + ", " + extraParams + ")");
            int callingUid = getCallingUid();

            if (!VendingPreferences.isLicensingEnabled(getApplicationContext())) {
                Log.d(TAG, "not checking license, as it is disabled by user");
                return;
            }

            Account[] accounts = accountManager.getAccountsByType(AuthConstants.DEFAULT_ACCOUNT_TYPE);
            PackageManager packageManager = getPackageManager();

            if (accounts.length == 0) {
                handleNoAccounts(packageName, packageManager);
            } else {
                checkLicenseV2(callingUid, packageName, packageManager, listener, extraParams, new LinkedList<>(Arrays.asList(accounts)));
            }
        }

        private void checkLicenseV2(int callingUid, String packageName, PackageManager packageManager,
                                    ILicenseV2ResultListener listener, Bundle extraParams,
                                    Queue<Account> remainingAccounts) throws RemoteException {
            new LicenseChecker.V2().checkLicense(
                remainingAccounts.poll(), accountManager, androidId, packageName, callingUid, packageManager, queue, Unit.INSTANCE,
                (responseCode, data) -> {
                    /*
                     * Suppress failures on V2. V2 is commonly used by free apps whose checker
                     * will not throw users out of the app if it never receives a response.
                     *
                     * This means that users who are signed in to a Google account will not
                     * get a worse experience in these apps than users that are not signed in.
                     */
                    if (responseCode == LICENSED) {
                        Bundle bundle = new Bundle();
                        bundle.putString(KEY_V2_RESULT_JWT, data);

                        listener.verifyLicense(responseCode, bundle);
                    } else if (!remainingAccounts.isEmpty()) {
                        checkLicenseV2(callingUid, packageName, packageManager, listener, extraParams, remainingAccounts);
                    } else {
                        Log.i(TAG, "Suppressed negative license result for package " + packageName);
                    }
                }
            );

        }

        private void handleNoAccounts(String packageName, PackageManager packageManager) {
            notificationRunnable.callerPackageName = packageName;
            try {
                Log.e(TAG, "not checking license, as user is not signed in");
                PackageInfo packageInfo = packageManager.getPackageInfo(packageName, 0);
                notificationRunnable.callerUid = packageInfo.applicationInfo.uid;
                notificationRunnable.callerAppName = packageManager.getApplicationLabel(packageInfo.applicationInfo);
                if (notificationRunnable.callerAppName == null) {
                    notificationRunnable.callerAppName = packageName;
                }
            } catch (PackageManager.NameNotFoundException e) {
                Log.e(TAG, "ignored license request, but package name " + packageName + " was not known!");
                notificationRunnable.callerAppName = packageName;
            }
            notificationRunnable.run();
        }
    };

    public IBinder onBind(Intent intent) {


        Cursor cursor = null;
        try {
            cursor = getContentResolver().query(
                PROFILE_PROVIDER, null, null, null, null
            );

            if (cursor == null || cursor.getColumnCount() != 2) {
                Log.e(TAG, "profile provider not available");
            } else {
                Map<String, String> profileData = new HashMap<>();
                while (cursor.moveToNext()) {
                    profileData.put(cursor.getString(0), cursor.getString(1));
                }
                ProfileManager.INSTANCE.applyProfileData(profileData);
            }

        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }

        try {
            cursor = getContentResolver().query(
                    CHECKIN_SETTINGS_PROVIDER, new String[] { "androidId" }, null, null, null
            );

            if (cursor != null) {
                cursor.moveToNext();
                androidId = Long.toHexString(cursor.getLong(0));
            }
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }

        queue = Volley.newRequestQueue(this);
        accountManager = AccountManager.get(this);
        notificationRunnable = new LicenseServiceNotificationRunnable(this);

        return mLicenseService;
    }


}

File: vending-app/src/main/java/com/android/vending/licensing/LicensingService.kt
/*
 * SPDX-FileCopyrightText: 2023 microG Project Team
 * SPDX-License-Identifier: Apache-2.0
 */
package com.android.vending.licensing

import android.accounts.Account
import android.accounts.AccountManager
import android.app.Service
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.os.IBinder
import android.os.RemoteException
import android.util.Log
import com.android.vending.VendingPreferences.isLicensingEnabled
import com.android.vending.VendingPreferences.isLicensingPurchaseFreeAppsEnabled
import kotlinx.coroutines.runBlocking
import org.microg.gms.auth.AuthConstants
import org.microg.gms.profile.ProfileManager.ensureInitialized
import org.microg.vending.billing.acquireFreeAppLicense
import org.microg.vending.billing.core.HttpClient

class LicensingService : Service() {
    private lateinit var accountManager: AccountManager
    private lateinit var androidId: String
    private lateinit var httpClient: HttpClient

    private val mLicenseService: ILicensingService.Stub = object : ILicensingService.Stub() {

        @Throws(RemoteException::class)
        override fun checkLicense(
            nonce: Long,
            packageName: String,
            listener: ILicenseResultListener
        ): Unit = runBlocking {
            Log.v(TAG, "checkLicense($nonce, $packageName)")

            val response = checkLicenseCommon(packageName, V1Parameters(nonce))

            /* If a license is found, it is now stored in `lastResponse`. Otherwise, it now contains
             * an error. In either case, we should send it to the application.
             */
            try {
                when (response) {
                    is V1Response -> listener.verifyLicense(response.result, response.signedData, response.signature)
                    is ErrorResponse -> listener.verifyLicense(response.result, null, null)
                    is V2Response -> Unit // should never happen
                    null -> Unit // no license check was performed at all
                }
            } catch (e: Exception) {
                Log.w(TAG, "Remote threw an exception while returning license result ${response}")
            }
        }

        @Throws(RemoteException::class)
        override fun checkLicenseV2(
            packageName: String,
            listener: ILicenseV2ResultListener,
            extraParams: Bundle
        ): Unit = runBlocking {
            Log.v(TAG, "checkLicenseV2($packageName, $extraParams)")

            val response = checkLicenseCommon(packageName, V2Parameters)

            /*
             * Suppress failures on V2. V2 is commonly used by free apps whose checker
             * will not throw users out of the app if it never receives a response.
             *
             * This means that users who are signed in to a Google account will not
             * get a worse experience in these apps than users that are not signed in.
             *
             * Normally, we would otherwise always send the response.
             */
            if (response?.result == LICENSED && response is V2Response) {
                val bundle = Bundle()
                bundle.putString(KEY_V2_RESULT_JWT, response.jwt)

                try {
                    listener.verifyLicense(response.result, bundle)
                } catch (e: Exception) {
                    Log.w(TAG, "Remote threw an exception while returning license result ${response}")
                }
            } else {
                Log.i(TAG, "Suppressed negative license result for package $packageName")
            }

        }

        /**
         * Checks for license on all accounts.
         *
         * @return `null` if no check is performed (for example, because the feature is disabled),
         * an instance of [LicenseResponse] otherwise.
         */
        suspend fun checkLicenseCommon(
            packageName: String,
            request: LicenseRequestParameters
        ): LicenseResponse? {
            val callingUid = getCallingUid()

            if (!isLicensingEnabled(this@LicensingService)) {
                Log.d(TAG, "not checking license, as it is disabled by user")
                return null
            }

            val packageInfo = try {
                packageManager.getPackageInfo(packageName, 0)
            } catch (e: PackageManager.NameNotFoundException) {
                Log.e(TAG,
                    "an app tried to request licenses for package $packageName, which does not exist"
                )
                return ErrorResponse(ERROR_INVALID_PACKAGE_NAME)
            }

            // Verify caller identity
            if (packageInfo.applicationInfo.uid != callingUid) {
                Log.e(
                    TAG,
                    "an app illegally tried to request licenses for another app (caller: $callingUid)"
                )
                return ErrorResponse(ERROR_NON_MATCHING_UID)
            }

            val accounts = accountManager.getAccountsByType(AuthConstants.DEFAULT_ACCOUNT_TYPE)
            val packageManager = packageManager

            lateinit var lastResponse: LicenseResponse
            if (accounts.isEmpty()) {
                handleNoAccounts(packageName, packageManager)
                return null
            } else for (account: Account in accounts) {

                lastResponse = httpClient.checkLicense(
                    account, accountManager, androidId, packageInfo, packageName, request
                )

                if (lastResponse.result == LICENSED) {
                    return lastResponse;
                }
            }

            // Attempt to acquire license if app is free ("auto-purchase")
            if (isLicensingPurchaseFreeAppsEnabled(this@LicensingService)) {
                val firstAccount = accounts[0]
                if (httpClient.acquireFreeAppLicense(
                        this@LicensingService,
                        firstAccount,
                        packageName
                    )
                ) {
                    lastResponse = httpClient.checkLicense(
                        firstAccount, accountManager, androidId, packageInfo, packageName, request
                    )
                }
            } else {
                Log.d(TAG, "Not auto-purchasing $packageName as it is disabled by the user")
            }

            return lastResponse
        }

        private fun handleNoAccounts(packageName: String, packageManager: PackageManager) {
            try {
                Log.e(TAG, "not checking license, as user is not signed in")

                packageManager.getPackageInfo(packageName, 0).let {
                    sendLicenseServiceNotification(
                        packageName,
                        packageManager.getApplicationLabel(it.applicationInfo),
                        it.applicationInfo.uid
                    )
                }

            } catch (e: PackageManager.NameNotFoundException) {
                Log.e(TAG, "ignored license request, but package name $packageName was not known!")
                // don't send sign in notification
            }
        }
    }

    override fun onBind(intent: Intent): IBinder {

        ensureInitialized(this)

        contentResolver.query(
            CHECKIN_SETTINGS_PROVIDER,
            arrayOf("androidId"),
            null,
            null,
            null
        ).use { cursor ->
            if (cursor != null) {
                cursor.moveToNext()
                androidId = java.lang.Long.toHexString(cursor.getLong(0))
            }
        }
        accountManager = AccountManager.get(this)
        httpClient = HttpClient()

        return mLicenseService
    }


    companion object {
        private const val TAG = "FakeLicenseService"
        private const val KEY_V2_RESULT_JWT = "LICENSE_DATA"

        private val CHECKIN_SETTINGS_PROVIDER: Uri =
            Uri.parse("content://com.google.android.gms.microg.settings/check-in")
    }
}
